### ISSUE TITLE
using connection DSN doesn't support sockets

### ISSUE THREAD

**Issue by thecodeassassin** [knex/knex#1568]
Hey all,

I want to create a connection with the following DSN:

```
mysql:/user:pass@/cloudsql/test:europe-west1:staging-database/dbname
```

Only the problem here is that parse-connection.js doesn't support sockets :(

**Comment by elhigu** [comment]
Looks like you are talking about connection string not supporting *UNIX* sockets... Also looks like you are missing on `/` from you connection string. To me it looks like it should be: 

```
mysql://user:pass@/cloudsql/test:europe-west1:staging-database/dbname
```

Anyways I'm pretty sure that parser doesn't work anyways for that kind of socket... did you try to pass connection parameters separately?

> connection: {
>     socketPath : '/path/to/socket.sock',
>     user     : 'your_database_user',
>     password : 'your_database_password',
>     database : 'myapp_test'
>   } 

Sounds like bug anyways... also passing "query parameters" e.g. for enabling ssl could be implemented too..

**Comment by thecodeassassin** [comment]
Yeah i meant unix sockets :) the connection string i posted is fine i just missed on / while copy pasting haha :)

I did not try that yet but that will require some overhaul of our application. Would still love to see this implemented :) Thank you for the fast response.

**Comment by mercmobily** [comment]
Thanks for reporting this issue about Unix socket support in connection strings, particularly for Google Cloud SQL connections.

**Good news! This has been resolved.** Since November 2019 (commit 07276bcb), Knex supports Unix sockets for MySQL/MariaDB via query parameters in the connection string.

**Solution**: You can now use the `socketPath` query parameter in your connection string:

```javascript
// Connection string with socketPath as query parameter
const db = knex('mysql://user:pass@localhost/dbname?socketPath=/cloudsql/test:europe-west1:staging-database');

// Or as a config object
const db = knex({
  client: 'mysql',
  connection: 'mysql://user:pass@localhost/dbname?socketPath=/cloudsql/test:europe-west1:staging-database'
});
```

**Alternative**: You can still use the connection object format:

```javascript
const db = knex({
  client: 'mysql',
  connection: {
    socketPath: '/cloudsql/test:europe-west1:staging-database',
    user: 'user',
    password: 'pass',
    database: 'dbname'
  }
});
```

**How it works**: 
- Since November 2019 (commit 07276bcb), Knex's connection string parser extracts ALL query parameters and adds them to the connection object
- The MySQL/MySQL2 drivers natively support `socketPath` in their connection configuration
- When you include `?socketPath=/path/to/socket` in your connection string, Knex passes it through to the driver

**Note**: The connection string must have a host (use `localhost` if connecting via socket). URLs without a host like `mysql://user:pass@/dbname` are interpreted as SQLite file paths.

This issue can be closed as it has been resolved since Knex v0.20.2 (November 2019).

**Related issues:**
- #3783 - Improved PostgreSQL socket support
- #2354 - Parsing postgres connection strings query parameters  
- #5145 - Query parameter handling in connection strings
- #611 - Initial socket path documentation
